### PR TITLE
feat!: register substratools function with decorator

### PR DESCRIPTION
**Pull request by ThibaultFy**
## Description

Register functions to substratools can be done with a decorator.

  ```py
  def my_function1:
    pass
  def my_function2:
    pass
  if __name__ == '__main__':
    tools.execute(my_function1, my_function2)
  ```

  become

  ```py
  @tools.register
  def my_function1:
    pass
  @tools.register
  def my_function2:
    pass
  if __name__ == '__main__':
    tools.execute()
  ```

## Companion PR:
- https://github.com/Substra/substra-tools/pull/65 (MAIN)
- https://github.com/Substra/substra/pull/310
- https://github.com/Substra/substrafl/pull/37
- https://github.com/Substra/substra-documentation/pull/202
- https://github.com/Substra/substra-tests/pull/225

### CODE REVIEW

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by Owlfred** [comment]
```
Usage: /e2e [options] [help]

/e2e may appear anywhere as long as it is on its own line

Options:
  --refs <value>                                         Extra refs (branch or tag) with format REPO=GIT_REF,REPO=GIT_REF.
  Supported repositories: substra-hlf-k8s, orchestrator, substra-backend, substra-frontend, substra-tools, substrafl, substra, substra-tests, substra-ci.
  Example: /e2e --refs substra-backend=some_branch,orchestrator=some_tag (default: {})
  --tests-to-run, --tests <tests-to-run>                 Comma-separated list of tests to run. Valid options: sdk,substrafl,frontend or NONE. (default: "sdk")
  --benchmarks-to-run, --benchmarks <benchmarks-to-run>  Comma-separated list of workflows tests to run. Valid options: mnist, camelyon or NONE. (default: "NONE")
  --orchestrator-mode, --mode <orchestrator-mode>        Comma-separated list of orchestrator modes to run tests for. Valid options: standalone,distributed (default: "standalone,distributed")
  -h, --help                                             display help for command
```

**Comment by ThibaultFy** [comment]
/e2e --refs substra=gt/decorator,substrafl=gt/decorator,substra-tests=gt/decorator,substra-tools=gt/decorator --tests NONE --benchmarks mnist,camelyon

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: substratools/function.py [hunk 1]
```diff
@@ -72,6 +73,45 @@ def _parser_add_default_arguments(parser):
     )
 
 
+class FunctionRegister:
+    """Class to create a decorator to register function in substratools. The functions are registered in the _functions
+    dictionary, with the function.__name__ as key.
+    Register a function in substratools means that this function can be access by the function.execute functions through
+    the --function-name given by the CLI."""
```
Comment (oleobal) [inline substratools/function.py:80]: typo
```suggestion
    Register a function in substratools means that this function can be accessed by the function.execute functions through
    the --function-name CLI argument."""
```

**Comment by oleobal** [comment]
Couldn't see anything unexpected, it looks good to me

**Review by Fabien-GELUS — APPROVED** [review]
LGTM!

**Review by oleobal — APPROVED** [review]

**Pull request merged**